hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergedReadHandle.java

```diff
@@ -0,0 +1,173 @@
+          .findFirst());
+    }
+    return Option.empty();
+  }
+
+  private Option<HoodieFileReader> getBaseFileReader(FileSlice fileSlice) throws IOException {
+    if (fileSlice.getBaseFile().isPresent()) {
+      return Option.of(createNewFileReader(fileSlice.getBaseFile().get()));
+    }
+    return Option.empty();
+  }
+
+  private HoodieMergedLogRecordScanner getLogRecordScanner(FileSlice fileSlice) {
+    List<String> logFilePaths = fileSlice.getLogFiles().map(l -> l.getPath().toString()).collect(toList());
+    return HoodieMergedLogRecordScanner.newBuilder()
+        .withFileSystem(hoodieTable.getMetaClient().getFs())
+        .withBasePath(hoodieTable.getMetaClient().getBasePathV2().toString())
+        .withLogFilePaths(logFilePaths)
+        .withReaderSchema(readerSchema)
+        .withLatestInstantTime(instantTime)
+        .withMaxMemorySizeInBytes(IOUtils.getMaxMemoryPerCompaction(hoodieTable.getTaskContextSupplier(), config))
+        .withReadBlocksLazily(config.getCompactionLazyBlockReadEnabled())
+        .withReverseReader(config.getCompactionReverseLogReadEnabled())
+        .withBufferSize(config.getMaxDFSStreamBufferSize())
+        .withSpillableMapBasePath(config.getSpillableMapBasePath())
+        .withPartition(fileSlice.getPartitionPath())
+        .withOptimizedLogBlocksScan(config.enableOptimizedLogBlocksScan())
+        .withDiskMapType(config.getCommonConfig().getSpillableDiskMapType())
+        .withBitCaskDiskMapCompressionEnabled(config.getCommonConfig().isBitCaskDiskMapCompressionEnabled())
+        .withRecordMerger(config.getRecordMerger())
+        .build();
+  }
+
+  private List<HoodieRecord<T>> doMergedRead(Option<HoodieFileReader> baseFileReaderOpt, HoodieMergedLogRecordScanner logRecordScanner) throws IOException {
+    List<HoodieRecord<T>> mergedRecords = new ArrayList<>();
+    Map<String, HoodieRecord> deltaRecordMap = logRecordScanner.getRecords();
+    Set<String> deltaRecordKeys = new HashSet<>(deltaRecordMap.keySet());
+
+    if (baseFileReaderOpt.isPresent()) {
+      HoodieFileReader baseFileReader = baseFileReaderOpt.get();
+      HoodieRecordMerger recordMerger = config.getRecordMerger();
+      ClosableIterator<HoodieRecord<T>> baseFileItr = baseFileReader.getRecordIterator(readerSchema);
+      HoodieTableConfig tableConfig = hoodieTable.getMetaClient().getTableConfig();
+      Option<Pair<String, String>> simpleKeyGenFieldsOpt =
+          tableConfig.populateMetaFields() ? Option.empty() : Option.of(Pair.of(tableConfig.getRecordKeyFieldProp(), tableConfig.getPartitionFieldProp()));
+      while (baseFileItr.hasNext()) {
+        HoodieRecord<T> record = baseFileItr.next().wrapIntoHoodieRecordPayloadWithParams(readerSchema,
+            config.getProps(), simpleKeyGenFieldsOpt, logRecordScanner.isWithOperationField(), logRecordScanner.getPartitionNameOverride(), false);
+        String key = record.getRecordKey();
+        if (deltaRecordMap.containsKey(key)) {
```

Comment:
I see this is pretty much what our MOR snapshot is doing. is there a way to re-use code? 
what incase we make some fixes to MOR snapshot, but missed to fix this code?

Reply:
nvm. I don't think we have any such code. but we need to remember if any fixes goes into MergeOnRead snapshot, we might have to fix here as well.

Reply:
this is better for filegroup reader to encapsulate.